ls1intum/Artemis
# Fix issues in the exam mode test runs

### Checklist

### Motivation and Context
We had a recurring issue where the user object sent from the client was stored on the database on a cascade operation. 

### Description
1. Remove the `cascade.Merge` on `Result.assessor`. 
2. Filter out the results from the submissions of student exams during conduction
3. Add a test_run flag for participations, adapt the `isTestRunParticipation` helper method. (In follow up PRs we will only use this flag to identify test run participations and their submissions)
3. Mark the test run submissions AFTER the instructor submits the test run & evaluate the quiz submission, instead of before.
4. Warn the instructor to delete the test runs before the conduction.
5. In case the instructor forgets to delete the testruns, delete all test runs automatically once the instructor assesses all unsubmitted exams.
6. Do not show second correction table for test runs (only the owner of the test runs can assess the submission once & override it)

### Steps for Testing
1. Create an exam with all exercise types
2. Participate in test runs  and submit. **IMPORTANT**: wait for the build result to return from bamboo!
3. Make sure that the submissions don't appear in the assessment dashboard 
4. Assess the test run submissions
5. Add complaints to the test run submissions
6. Review the complaints
7. View the summary page of the test run, make sure that the results appear correctly. 
8. Make sure that the test runs don't appear in the exam scores
9. Start a test run and do not submit
10. Participate in the exam as a student
11. Assess all unsubmitted exams (button in the student exams tab)
12. Verify that all test runs are deleted.

13. Assess the exam submissions
14. Override with an instructor account
15. Complain about the submissions
17. Review complaints 
18. view scores


### Test Coverage
`StudentExamService` 90%
`ExamService` 91%
`ParticipationService` 72% -> After a followup rework of identifying test run participations will increase to > 80%

### Screenshots
![image](https://user-images.githubusercontent.com/15110960/105695689-a784be80-5f02-11eb-8e76-8487be47e37e.png)
